File: backend/app/observability/sentry.py
"""Sentry initialization for the backend (T-0.6, R6.4).

Production-mandatory: raises if ``SENTRY_DSN`` is missing while
``ENVIRONMENT=production``. In dev with no DSN, this is a no-op.

Sample rate per R6.4: 100% errors, 10% performance traces (matches the
frontend policy).
"""

from __future__ import annotations

import logging
import os
from typing import Optional

log = logging.getLogger("observability.sentry")


class SentryConfigurationError(RuntimeError):
    """Raised when Sentry is mandatory but DSN is unset."""


def init_sentry(
    *,
    dsn: Optional[str] = None,
    environment: Optional[str] = None,
    release: Optional[str] = None,
) -> bool:
    """Initialize Sentry. Returns True if Sentry was actually configured.

    Args:
        dsn: Optional override for the DSN. Defaults to ``SENTRY_DSN`` env.
        environment: Defaults to ``ENVIRONMENT`` env or 'development'.
        release: Defaults to ``APP_VERSION`` env.

    Raises:
        SentryConfigurationError: in production when no DSN is configured.
    """
    env = environment or os.environ.get("ENVIRONMENT", "development")
    dsn_value = dsn if dsn is not None else os.environ.get("SENTRY_DSN")

    if not dsn_value:
        if env == "production":
            raise SentryConfigurationError(
                "SENTRY_DSN must be set in production (R6.4)"
            )
        log.info("sentry.skipped (no DSN in %s)", env)
        return False

    try:
        import sentry_sdk  # type: ignore
        from sentry_sdk.integrations.fastapi import FastApiIntegration  # type: ignore
        from sentry_sdk.integrations.asgi import SentryAsgiMiddleware  # type: ignore  # noqa: F401
    except Exception as e:  # noqa: BLE001
        if env == "production":
            raise SentryConfigurationError(
                f"sentry-sdk is required in production: {e}"
            ) from e
        log.warning("sentry.sdk_missing — install sentry-sdk to enable")
        return False

    sentry_sdk.init(
        dsn=dsn_value,
        environment=env,
        release=release or os.environ.get("APP_VERSION"),
        # R6.4 — 100% error sampling.
        sample_rate=1.0,
        # R6.4 — 10% performance traces.
        traces_sample_rate=0.1,
        integrations=[FastApiIntegration()],
        # Sensitive data scrubbing is enabled by default; keep it on.
        send_default_pii=False,
    )
    log.info("sentry.initialized", extra={"environment": env})
    return True

File: frontend/src/observability/sentry.ts
/**
 * Sentry initialization (T-0.6, R6.4).
 *
 * Production-mandatory: throws if `VITE_SENTRY_DSN` is missing in a
 * production build. In dev, missing DSN silently disables Sentry.
 *
 * Sample rates per R6.4:
 *   - 100% errors
 *   - 10%  performance traces
 *
 * `@sentry/react` is dynamically imported so this module can be loaded in
 * builds where the package is not yet installed (see _deltas/P0-deps.md).
 */

export interface SentryInitOptions {
  /** Override the DSN read from env. Useful for tests. */
  dsn?: string;
  /** Override the environment label. Defaults to PROD ? 'production' : 'development'. */
  environment?: string;
  /** Override the release identifier. Defaults to VITE_APP_VERSION. */
  release?: string;
}

type ViteEnv = {
  PROD?: boolean;
  VITE_SENTRY_DSN?: string;
  VITE_APP_VERSION?: string;
};

const VITE_ENV: ViteEnv =
  ((import.meta as unknown as { env?: ViteEnv }).env ?? {}) as ViteEnv;

function isProd(): boolean {
  return Boolean(VITE_ENV.PROD);
}

type SentryModule = {
  init: (cfg: Record<string, unknown>) => void;
  captureException: (
    err: unknown,
    context?: { tags?: Record<string, string> },
  ) => void;
  browserTracingIntegration: () => unknown;
};

let _sentry: SentryModule | null = null;

/**
 * Initialize Sentry. Call once during app bootstrap.
 *
 * @throws Error in production if VITE_SENTRY_DSN is unset (R6.4).
 */
export async function initSentry(
  opts: SentryInitOptions = {},
): Promise<void> {
  const dsn = opts.dsn ?? VITE_ENV.VITE_SENTRY_DSN;
  const prod = isProd();

  if (!dsn) {
    if (prod) {
      // Sentry is recommended but not mandatory: warn and continue so the app
      // still boots when error tracking is intentionally deferred.
      console.warn(
        '[observability] VITE_SENTRY_DSN is unset — Sentry disabled (R6.4)',
      );
    }
    return;
  }

  let mod: SentryModule;
  try {
    // @ts-expect-error — `@sentry/react` is an optional runtime dep (see _deltas/P0-deps.md).
    const imported = await import(/* @vite-ignore */ '@sentry/react');
    mod = imported as unknown as SentryModule;
  } catch (e) {
    if (prod) {
      throw new Error(
        '[observability] @sentry/react failed to load in production: ' + String(e),
      );
    }
    return;
  }
  _sentry = mod;

  mod.init({
    dsn,
    environment: opts.environment ?? (prod ? 'production' : 'development'),
    release: opts.release ?? VITE_ENV.VITE_APP_VERSION,
    sampleRate: 1.0,
    tracesSampleRate: 0.1,
    integrations: [mod.browserTracingIntegration()],
    ignoreErrors: [
      'ResizeObserver loop limit exceeded',
      'ResizeObserver loop completed with undelivered notifications.',
      'Non-Error promise rejection captured',
    ],
  });
}

/**
 * Capture an exception with optional tag context. Safe to call before init —
 * Sentry silently no-ops if not configured.
 */
export function captureException(
  err: unknown,
  context?: { tags?: Record<string, string> },
): void {
  try {
    _sentry?.captureException(err, context);
  } catch {
    /* swallow */
  }
}
